src/crypto/CHIPCryptoPALOpenSSL.cpp

```diff
@@ -1665,5 +1666,60 @@ CHIP_ERROR ExtractPubkeyFromX509Cert(const ByteSpan & certificate, Crypto::P256P
     return err;
 }
 
+CHIP_ERROR ExtractAKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan & akid)
+{
+    CHIP_ERROR err                       = CHIP_NO_ERROR;
+    X509 * x509certificate               = nullptr;
+    const unsigned char * pCertificate   = certificate.data();
+    const unsigned char ** ppCertificate = &pCertificate;
+    const ASN1_OCTET_STRING * akidString = nullptr;
+
+    x509certificate = d2i_X509(NULL, ppCertificate, static_cast<long>(certificate.size()));
+    VerifyOrExit(x509certificate != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    akidString = X509_get0_authority_key_id(x509certificate);
+
+    VerifyOrExit(akidString->length == static_cast<int>(akid.size()), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    for (int i = 0; i < akidString->length; ++i)
+    {
+        akid.data()[i] = akidString->data[i];
+    }
+
+exit:
+    X509_free(x509certificate);
+
+    return err;
+}
+
+CHIP_ERROR ExtractVIDFromX509Cert(const ByteSpan & certificate, uint16_t & vid)
+{
+    CHIP_ERROR err                        = CHIP_NO_ERROR;
+    X509 * x509certificate                = nullptr;
+    const unsigned char * pCertificate    = certificate.data();
+    const unsigned char ** ppCertificate  = &pCertificate;
+    char * subj                           = nullptr;
+    char * vidSubj                        = nullptr;
+    static constexpr char vidNeedle[]     = "1.3.6.1.4.1.37244.2.1=";
+    static constexpr size_t vidNeedleSize = sizeof(vidNeedle) - 1;
+
+    x509certificate = d2i_X509(NULL, ppCertificate, static_cast<long>(certificate.size()));
+    VerifyOrExit(x509certificate != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    subj = X509_NAME_oneline(X509_get_subject_name(x509certificate), NULL, 0);
+    VerifyOrExit(subj != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    vidSubj = strstr(subj, vidNeedle);
+    VerifyOrExit(vidSubj != nullptr, err = CHIP_ERROR_WRONG_CERT_TYPE);
+
+    vid = static_cast<uint16_t>(strtol(vidSubj + vidNeedleSize, NULL, 16));
```

Comment:
This needs to be `strtoul`. This also does not validate format. It's probably fine not to check for exact width, but we should check for error (e.g. not pass NULL for second arg). Overall, this likely will fail if the subject does not *end* with a VID, since the `strtoul` will see a subject separator.

Ex: `CN=Matter Test DAC 000C,1.3.6.1.4.1.37244.2.1=FFF1,1.3.6.1.4.1.37244.2.2=8001`

The method won't work there.

Also, docs for openssl state:
```
NOTES
The functions X509_NAME_oneline() and X509_NAME_print() are legacy functions which produce a non standard output form, they don't handle multi character fields and have various quirks and inconsistencies. Their use is strongly discouraged in new applications.
```

Please find an alternate method to extract VID

Reply:
@emargolis Aren't there other methods already in tree to extract RDNs from x.509 subject fields?